### PR TITLE
fix(flox-bash): make  `selectAttrPath` more resilient

### PR DESCRIPTION
**Pull request by ysndr**
## Proposed Changes

* exit if selection required without tty

Let `selectAttrPath` raise an error if more than one option exist but we can't prompt users.
 
* exit if choosing a package 

Fail `selectAttrPath` if `gum choose` command failed.

This can be either due to an error in gum or (more importantly) an explicit interrupt by the user.

This change was originally motivated by providing better errors for `flox activate`, but since `selectAttrPath` is used by several commands all of them will benefit from it.

### CODE REVIEW

File: flox-bash/lib/utils.sh [hunk 1]
```diff
@@ -1356,8 +1356,15 @@ function selectAttrPath() {
 	elif [ ${#attrPaths[@]} -eq 1 ]; then
 		echo "${attrPaths[0]}"
 	else
-		warn "Select package for flox $subcommand"
-		attrPath=$($_gum choose ${attrPaths[*]})
+		# don't attempt to prompt if not interactive
+		if [ ! -t 0 ] || [ ! -t 2 ]; then
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply flox-bash/lib/utils.sh:1360]: Please let's fix $interactive and use it. Patch forthcoming.

**Review by limeytexan — APPROVED** [review]
LGTM!

File: flox-bash/lib/bootstrap.sh [hunk 2]
```diff
@@ -25,7 +25,7 @@ function bootstrap() {
 	floxUserMetaRegistry get floxClientUUID >/dev/null || \
 		floxUserMetaRegistry set floxClientUUID $($_uuid)
 	floxClientUUID=$(floxUserMetaRegistry get floxClientUUID)
-	if [ -t 1 ]; then
```
Comment (ysndr) [inline flox-bash/lib/bootstrap.sh:28]: I think that breaks the "non-interactive" `flox activate`, as in:

`$ eval $(flox activate)` 

run in an interactive shell session (that includes `.bashrc/.zshrc`).

IMO there is a difference between an "interactive" that changes output and formatting (i.e. how git/ls/cargo/nix output changes when run in ci) and an "interactive" whare a human _could_ answer to prompts.

Those are not necessarily the same: In `flox activate` we do want to detect if the output is collected but we still want to be able to prompt users.

**Review by ysndr — COMMENTED** [review]
May require revisiting interactive for `activate`

File: flox-bash/lib/bootstrap.sh (hunk 2 above)
Comment (ysndr) [reply flox-bash/lib/bootstrap.sh:28]: would you mind checking if [55fe699](https://github.com/flox/flox/pull/112/commits/55fe699d1726f862a72792489040476b6b0d0bba) changes the right references to $interactive?

There is one left that seems to fall in the "logging" kind of interaction so i left that the same

Comment (limeytexan) [reply flox-bash/lib/bootstrap.sh:28]: I think so .. and I was working on the same patch on this end! My version was a little more explicit though by creating a new `$spawnMode` variable - I think the logic is more clear and will prevent us from conflating these concepts again by giving it a name. Can you have a look at 741ac9be0d0755760f494df4cc6f1fc2e79b1a79 to see if that looks more clear?

**Review by limeytexan — APPROVED** [review]

File: flox-bash/lib/commands/activate.sh [hunk 3]
```diff
@@ -251,7 +252,7 @@ function floxActivate() {
 	# https://discourse.floxdev.com/t/losing-part-of-my-shell-environment-when-using-flox-develop/556/2
 	if [[ -x /usr/libexec/path_helper ]] && [[ "$PATH" =~ ^/usr/local/bin: ]]
 	then
-		if [[ "${#cmdArgs[@]}" -eq 0 ]] && [[ "$interactive" -eq 0 ]]; then
+		if [[ "${#cmdArgs[@]}" -eq 0 ]] && [[ "$spawnMode" -eq 0 ]]; then
```
Comment (ysndr) [inline flox-bash/lib/commands/activate.sh:255]:
```suggestion
		if [[ "${#cmdArgs[@]}" -eq 0 ]] && [[ "$spawnMode" -eq 1 ]]; then
```

**Comment by ysndr** [comment]
@limeytexan have you seen my comment on your previous commit and the suggestion i made on the PR? 

I think we swapped the logic there a bit - might be on me by shameless adding an ! in the condition in my commit earlier 😉

**Comment by limeytexan** [comment]
> @limeytexan have you seen my comment on your previous commit and the suggestion i made on the PR?
> 
> I think we swapped the logic there a bit - might be on me by shameless adding an ! in the condition in my commit earlier 😉

The polarity of `$spawnMode` matches that of the `$interactive` reference it replaces, and you'll see none of the other uses of the variable flipped polarity, so no I don't believe this is wrong. That said, you do highlight that the comment could more clearly explain the situation that this is meant to address: if you run `tmux` from an already-activated environment the subsequent shell is launched with stdout directed to something other than a TTY, and then the darwin `path_helper` rearranges the PATH. The reproducer is simple: add `eval "$(flox activate)` to your .zshrc, then run `flox activate; tmux` and then confirm that `/usr/local/bin` is not first in the `$PATH`.

**Comment by ysndr** [comment]
I'd still want to rename `spawnMode` eventually, It's very specific for a global and our logic seems to check whether its off which all just ist quite obvious. No need to block on that bikeshed a better name now..

**Pull request merged**